crates/rattler_networking/src/mirror_middleware.rs

```diff
@@ -0,0 +1,238 @@
+
+    failures: AtomicUsize,
+
+    settings: MirrorSettings,
+}
+
+impl MirrorState {
+    pub fn add_failure(&self) {
+        self.failures.fetch_add(1, atomic::Ordering::Relaxed);
+    }
+}
+
+/// Middleware to handle mirrors
+pub struct MirrorMiddleware {
+    mirror_map: HashMap<String, Vec<MirrorState>>,
+}
+
+impl MirrorMiddleware {
+    /// Create a new `MirrorMiddleware` from a map of mirrors
+    pub fn from_map(map: HashMap<String, Vec<String>>) -> Self {
+        let mirror_map = map
+            .into_iter()
+            .map(|(k, v)| {
+                let v = v
+                    .into_iter()
+                    .map(|url| {
+                        let url = if url.ends_with('/') {
+                            url
+                        } else {
+                            format!("{url}/")
+                        };
+                        MirrorState {
+                            url: Url::parse(&url).unwrap(),
+                            failures: AtomicUsize::new(0),
+                            settings: MirrorSettings {
+                                no_zstd: false,
+                                no_bz2: false,
+                                no_gz: false,
+                                max_failures: Some(3),
+                            },
+                        }
+                    })
+                    .collect();
+                (k, v)
+            })
+            .collect();
+
+        Self { mirror_map }
+    }
+}
+
+fn select_mirror(mirrors: &[MirrorState]) -> &MirrorState {
+    let mut min_failures = usize::MAX;
+    let mut min_failures_index = 0;
+
+    for (i, mirror) in mirrors.iter().enumerate() {
+        let failures = mirror.failures.load(atomic::Ordering::Relaxed);
+        if failures < min_failures {
+            min_failures = failures;
+            min_failures_index = i;
+        }
+    }
+
+    &mirrors[min_failures_index]
+}
+
+#[async_trait::async_trait]
+impl Middleware for MirrorMiddleware {
+    async fn handle(
+        &self,
+        mut req: Request,
+        extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> Result<Response> {
+        let url_str = req.url().to_string();
+
+        for (key, mirrors) in self.mirror_map.iter() {
+            if let Some(url_rest) = url_str.strip_prefix(key) {
+                let url_rest = url_rest.trim_start_matches('/');
+                // replace the key with the mirror
+                let selected_mirror = select_mirror(mirrors);
+                let selected_url = selected_mirror.url.join(url_rest).unwrap();
+                *req.url_mut() = selected_url;
+                let res = next.run(req, extensions).await;
+
+                // record a failure if the request failed so we can avoid the mirror in the future
+                match res.as_ref() {
+                    Ok(res) if res.status().is_server_error() => selected_mirror.add_failure(),
+                    Err(_) => selected_mirror.add_failure(),
```

Comment:
Shouldnt we retry with another mirror?

Reply:
Retry is handled by another middleware that should come before. I think that's better, no?